### PR TITLE
feat(replacer): allow code to be evaluated

### PR DESCRIPTION
**Pull request by P0lip**
This would close https://github.com/stoplightio/spectral/issues/1339

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

**Additional context**

I'm not quite sure whether we'll want to have this in in the end.
While, I like the new syntax here, we already have these sort-of pipe operators, which are somewhat redundant as of this change. I'd rather avoid any fragmentation here, and go with a single syntax, but this would be a breaking change.

expression-eval is used, as it's slightly safer than running plain eval, as the set of operations available is strictly limited.

EDIT:
I decided to get rid of the "pipe syntax" and use functions everywhere where necessary, as this gives us more flexibility.
Luckily, we didn't document that syntax, therefore it's not a breaking change in the end.

### CODE REVIEW

File: src/utils/replacer.ts [hunk 1]
```diff
@@ -1,35 +1,46 @@
...
+            eval(parse(identifier), {
+              ...Object.entries(this.functions).reduce((fns, [name, fn]) => {
+                fns[name] = fn.bind(values);
+                return fns;
+              }, {}),
+              ...values,
+            }),
+          );
+        } catch {
+          return '';
```
[the remark this answers is not in the document]
Comment (P0lip) [reply src/utils/replacer.ts:36]: Yeah, I was thinking about it.
`console.warn`?

[the remark this answers is not in the document]
Comment (nulltoken) [reply src/utils/replacer.ts:37]: 🆒

File: src/utils/__tests__/replacer.test.ts [hunk 2]
```diff
@@ -53,16 +53,42 @@ describe('Replacer', () => {
...
+    const template = '#{{printPath()}}';
 
     expect(
       replacer.print(template, {
         path: ['foo', 'bar', '/a'],
       }),
     ).toEqual('foo.bar./a');
   });
+
+  it('handles exceptions thrown during evaluation', () => {
```
[the remark this answers is not in the document]
Comment (nulltoken) [reply src/utils/__tests__/replacer.test.ts]: @P0lip So you don't like my proposal, do you? ;-)

File: src/utils/__tests__/replacer.test.ts [hunk 3]
```diff
@@ -53,16 +57,47 @@ describe('Replacer', () => {
...
+    const template = '#{{printPath()}}';
 
     expect(
       replacer.print(template, {
         path: ['foo', 'bar', '/a'],
       }),
     ).toEqual('foo.bar./a');
   });
+
+  it('handles and prints out exceptions thrown during evaluation', () => {
```
Comment (P0lip) [inline src/utils/__tests__/replacer.test.ts:88]:
```suggestion
  it('ignores and prints out exceptions thrown during evaluation', () => {
```
@nulltoken looks good? :grin:

**Review by nulltoken — APPROVED** [review]

**Pull request merged**